lib/semantics/resolve-names.cc

```diff
@@ -420,18 +425,127 @@ class SubprogramVisitor : public InterfaceVisitor {
   // Function result name from parser::Suffix, if any.
   const parser::Name *funcResultName_{nullptr};
 
-  bool BeginSubprogram(const parser::Name &,
+  bool BeginSubprogram(const parser::Name &, Symbol::Flag,
       const std::optional<parser::InternalSubprogramPart> &);
   void EndSubprogram();
   // Create a subprogram symbol in the current scope and push a new scope.
-  Symbol &PushSubprogramScope(const parser::Name &);
+  Symbol &PushSubprogramScope(const parser::Name &, Symbol::Flag);
   Symbol *GetSpecificFromGeneric(const parser::Name &);
 };
 
+class DeclarationVisitor : public ArraySpecVisitor,
+                           public virtual ScopeHandler {
+public:
+  using ArraySpecVisitor::Post;
+  using ArraySpecVisitor::Pre;
+
+  void Post(const parser::EntityDecl &);
+  void Post(const parser::ObjectDecl &);
+  bool Pre(const parser::AsynchronousStmt &);
+  bool Pre(const parser::ContiguousStmt &);
+  bool Pre(const parser::ExternalStmt &);
+  bool Pre(const parser::IntrinsicStmt &);
+  bool Pre(const parser::OptionalStmt &);
+  bool Pre(const parser::ProtectedStmt &);
+  bool Pre(const parser::ValueStmt &);
+  bool Pre(const parser::VolatileStmt &);
+  bool Pre(const parser::AllocatableStmt &) {
+    objectDeclAttr_ = Attr::ALLOCATABLE;
+    return true;
+  }
+  void Post(const parser::AllocatableStmt &) { objectDeclAttr_ = std::nullopt; }
+  bool Pre(const parser::TargetStmt &x) {
+    objectDeclAttr_ = Attr::TARGET;
+    return true;
+  }
+  void Post(const parser::TargetStmt &) { objectDeclAttr_ = std::nullopt; }
+  void Post(const parser::DimensionStmt::Declaration &);
+  bool Pre(const parser::TypeDeclarationStmt &) { return BeginDecl(); }
+  void Post(const parser::TypeDeclarationStmt &) { EndDecl(); }
+  bool Pre(const parser::DerivedTypeDef &x);
+  void Post(const parser::DerivedTypeDef &x);
+  bool Pre(const parser::DerivedTypeStmt &x);
+  void Post(const parser::DerivedTypeStmt &x);
+  bool Pre(const parser::TypeAttrSpec::Extends &x);
+  bool Pre(const parser::PrivateStmt &x);
+  bool Pre(const parser::SequenceStmt &x);
+  bool Pre(const parser::ComponentDefStmt &) { return BeginDecl(); }
+  void Post(const parser::ComponentDefStmt &) { EndDecl(); }
+  void Post(const parser::ComponentDecl &x);
+  bool Pre(const parser::ProcedureDeclarationStmt &);
+  void Post(const parser::ProcedureDeclarationStmt &);
+  bool Pre(const parser::ProcComponentDefStmt &);
+  void Post(const parser::ProcComponentDefStmt &);
+  void Post(const parser::ProcInterface &x);
+  void Post(const parser::ProcDecl &x);
+  bool Pre(const parser::FinalProcedureStmt &x);
+
+protected:
+  bool BeginDecl();
+  void EndDecl();
+
+private:
+  // The attribute corresponding to the statement containing an ObjectDecl
```

Comment:
Minor: Data members (apart from constants) typically come after member functions in each accessibility section in a class definition.  I don't think that this point is in the style guide directly, though.